feat: support password command for YubiKey challenge-response

majiayu000 opened the pull request:
## Summary

This updates the original approach for #77 based on maintainer feedback.

Instead of toggling behavior with `TOTP_YUBIKEY` or treating a Yubico OTP as a reusable password, this PR adds a global `--password-command` flag. The command's stdout is used as the storage password.

That allows a real YubiKey HMAC-SHA1 challenge-response setup, for example:

```shell
totp-cli --password-command 'ykman otp calculate 2 746f74702d636c69' list
```

Because the challenge is stable, the YubiKey response is stable too. The README now explicitly warns not to use Yubico OTP/HOTP keyboard output as the password source because those values change on every touch.

## Changes

- Added global `--password-command`
- Added command-output password handling before command execution
- Replaced the old `TOTP_YUBIKEY` documentation with challenge-response focused documentation
- Added unit coverage for password command handling
- Applied required formatter output to existing OTP tests so `make lint` passes

## Testing

- `go test ./...`
- `make lint`

majiayu000 committed:
Add password command support

Add a global --password-command flag that reads the storage password from a user-provided command's stdout. This supports YubiKey HMAC-SHA1 challenge-response through tools such as ykman otp calculate without treating dynamic OTP output as a reusable password.

Fixes #77

Signed-off-by: majiayu000 <1835304752@qq.com>

yitsushi commented on internal/storage/filebackend.go:
```diff
@@ -65,8 +65,15 @@ func (s *FileBackend) Prepare() error {
-			if password, err = term.Hidden("Password:"); err != nil {
-				return BackendError{Message: err.Error()}
+			// Check if YubiKey mode is enabled
+			if os.Getenv("TOTP_YUBIKEY") == "1" || os.Getenv("TOTP_YUBIKEY") == "true" {
```
I don't think environment variables are the best to toggle functionality. They can provide information to automate things for example in a CI environment, but they should not control the flow itself.

The application takes in `TOTP_PASS` for password when used in CI, `TOTP_CLI_CREDENTIAL_FILE` to relocate the file permanently where it's stored, but it doesn't change how the application works, and TOTP_TOKEN for instant actions, that's still just a input to make it easier to work with CI.

Everything else is controlled by flags only like `follow`, `clear`, `prefix`, and so on. As multiple commands can use them and they should always behave the same way, they are stored in `internal/cmd/flags.go` so it can be reused on multiple `cli.Command{}`.

yitsushi reviewed (changes requested):
Sorry, I didn't see that PR earlier.

About the PR itself:
I can see that's a generated code which is not a problem, but I don't think it does what you want it to do. The whole point of a yubikey button is to not have a static password.

Please test before submitting a PR. This PR only changes the prompt, and still uses a static string as password. That simply wouldn't work with yubikey. When you touch the key two times, you will get a different string the second time.

I don't mind fully generated PRs as long as someone made sure it's working and safe. The burden of careful review shouldn't be pushed on the maintainer only. Whoever makes a contribution they should make sure they understand what they are filing, how it works, and make sure it works. Only after that the maintainer can review what's in there.

I'm happy if someone implements a yubikey feature, but this PR is not implementing that feature... or I'm missing something, and in that case, please tell me or show me it works and why.

Thank you :)

majiayu000 commented:
Thanks for the review. You were right: the previous version only changed the prompt and incorrectly described dynamic YubiKey OTP output as something that could be reused as an encryption password.

I rewrote the PR around a global `--password-command` flag instead:

- removed the `TOTP_YUBIKEY` behavior switch
- stopped documenting Yubico OTP/HOTP keyboard output as supported for this use case
- documented YubiKey HMAC-SHA1 challenge-response via `ykman otp calculate` with a stable challenge
- added unit coverage for the command-output password path
- verified with `go test ./...` and `make lint`

The intent is that totp-cli receives a stable password from a user-provided command, while the YubiKey secret stays on the device.

yitsushi reviewed (commented):
I'm sorry I really tried to understand what's the purpose of this PR in this form, but I couldn't get it.

why is it any better than just pipe the output of any program's output as password? This is already documented here: https://github.com/yitsushi/totp-cli#about-the-password
